File: tests-utils/src/main/java/tech/beshu/ror/utils/httpclient/RestClient.java
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.utils.httpclient;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import org.apache.http.Header;
import org.apache.http.auth.UsernamePasswordCredentials;
import org.apache.http.client.HttpClient;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.config.SocketConfig;
import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
import org.apache.http.conn.ssl.TrustSelfSignedStrategy;
import org.apache.http.impl.auth.BasicScheme;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.impl.client.StandardHttpRequestRetryHandler;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.ssl.SSLContextBuilder;
import tech.beshu.ror.utils.misc.Tuple;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class RestClient {

  private final CloseableHttpClient underlying;
  private final String host;
  private final int port;
  private final boolean ssl;

  public RestClient(boolean ssl, String host, int port) {
    this.ssl = ssl;
    this.host = host;
    this.port = port;
    this.underlying = createUnderlyingClient(Optional.empty());
  }

  public RestClient(boolean ssl, String host, int port, Optional<Tuple<String, String>> basicAuth, Header... headers) {
    this.ssl = ssl;
    this.host = host;
    this.port = port;
    this.underlying = createUnderlyingClient(basicAuth, headers);
  }

  private CloseableHttpClient createUnderlyingClient(Optional<Tuple<String, String>> basicAuth, Header... headers) {
    HttpClientBuilder builder = null;

    if (ssl) {
      try {
        SSLContextBuilder sslCtxBuilder = new SSLContextBuilder();
        sslCtxBuilder.loadTrustMaterial(null, new TrustSelfSignedStrategy());
        SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(sslCtxBuilder.build(), SSLConnectionSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER);
        builder = HttpClients.custom().setSSLSocketFactory(sslsf);
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
    else {
      builder = HttpClientBuilder.create();
    }

    // Common ops
    if (basicAuth.isPresent()) {
      System.out.println("SETTING CREDENTIALS " + (basicAuth.isPresent() ? (basicAuth.get().v1() + ":" + basicAuth.get().v2()) : ""));
      Header auth = BasicScheme.authenticate(
        new UsernamePasswordCredentials(basicAuth.get().v1(), basicAuth.get().v2()), "UTF-8", false);

      Header[] tmp = Arrays.copyOf(headers, headers.length + 1);
      tmp[tmp.length - 1] = auth;
      headers = tmp;
    }

    int timeout = 5;
    builder
      .setRetryHandler(new StandardHttpRequestRetryHandler(3, true))
      .setDefaultHeaders(Lists.newArrayList(headers))
      .setDefaultSocketConfig(SocketConfig.custom().build())
      .setDefaultRequestConfig(
          RequestConfig.custom()
              .setConnectTimeout(timeout * 1000)
              .setConnectionRequestTimeout(timeout * 1000)
              .setSocketTimeout(timeout * 1000).build()
      );

    return builder.build();
  }

  public HttpClient getUnderlyingClient() {
    return underlying;
  }

  public String getHost() {
    return host;
  }

  public int getPort() {
    return port;
  }

  public URI from(String path) throws URISyntaxException {
    return from(path, Maps.newHashMap());
  }

  public URI from(String path, Map<String, String> queryParams) throws URISyntaxException {
    URIBuilder uriBuilder = new URIBuilder();
    uriBuilder
        .setScheme(ssl ? "https" : "http")
        .setHost(host)
        .setPort(port)
        .setPath(("/" + path + "/").replaceAll("//", "/"));
    if (!queryParams.isEmpty()) {
      uriBuilder.setParameters(
        queryParams.entrySet().stream()
          .map(e -> new BasicNameValuePair(e.getKey(), e.getValue()))
          .collect(Collectors.toList())
      );
    }
    return uriBuilder.build();

  }

  public CloseableHttpResponse execute(HttpUriRequest req) throws IOException {
    return underlying.execute(req);
  }

}

File: tests-utils/src/main/scala/tech/beshu/ror/utils/httpclient/RestClient.scala
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.utils.httpclient

import monix.eval.Task
import monix.execution.Scheduler.Implicits.global
import org.apache.http.auth.UsernamePasswordCredentials
import org.apache.http.client.config.RequestConfig
import org.apache.http.client.methods.{CloseableHttpResponse, HttpUriRequest}
import org.apache.http.client.utils.URIBuilder
import org.apache.http.config.SocketConfig
import org.apache.http.conn.HttpHostConnectException
import org.apache.http.conn.ssl.SSLConnectionSocketFactory
import org.apache.http.conn.ssl.NoopHostnameVerifier
import org.apache.http.impl.auth.BasicScheme
import org.apache.http.impl.client.{HttpClientBuilder, HttpClients, StandardHttpRequestRetryHandler}
import org.apache.http.message.BasicHeader
import org.apache.http.ssl.SSLContextBuilder
import org.apache.http.util.EntityUtils
import org.apache.http.{Header, HttpResponse}

import java.net.URI
import scala.concurrent.duration._
import scala.collection.JavaConverters._
import scala.language.postfixOps
import scala.util.Try

class RestClient(ssl: Boolean,
                 host: String,
                 port: Int,
                 basicAuth: Option[(String, String)],
                 headers: Header*) {

  def this(ssl: Boolean,
           host: String,
           port: Int) = {
    this(ssl, host, port, None)
  }

  private val underlying = createUnderlyingClient(
    headers ++ basicAuth.map { case (u, p) => createBasicAuthHeader(u, p) }
  )

  def from(path: String, queryParams: Map[String, String]): URI = buildUri(path, queryParams)

  def from(path: String): URI = buildUri(path, Map.empty)

  def execute(req: HttpUriRequest): CloseableHttpResponse = {
    Task(underlying.execute(req))
      .onErrorRestartLoop(0) {
        case (ex: HttpHostConnectException, 10, _) =>
          Task.raiseError(new Exception("Cannot connect to the host", ex))
        case (_: HttpHostConnectException, num, retry) =>
          retry(num + 1).delayExecution(1 second)
        case (ex, _, _) =>
          Task.raiseError(ex)
      }
      .runSyncUnsafe()
  }

  private def createUnderlyingClient(headers: Iterable[Header]) = {
    val timeout = readTimeout()
    val builder =
      if (ssl) withSsl(HttpClients.custom())
      else HttpClientBuilder.create()
    builder
      .setRetryHandler(new StandardHttpRequestRetryHandler(3, true))
      .setDefaultHeaders(headers.toSeq.asJava)
      .setDefaultSocketConfig(SocketConfig.custom().build())
      .setDefaultRequestConfig(
        RequestConfig.custom()
          .setConnectTimeout(timeout * 1000)
          .setConnectionRequestTimeout(timeout * 1000)
          .setSocketTimeout(timeout * 1000).build()
      )
      .build()
  }

  private def withSsl(builder: HttpClientBuilder): HttpClientBuilder = {
    val sslCtxBuilder = new SSLContextBuilder();
    sslCtxBuilder.loadTrustMaterial(null, new TrustAllCertificatesStrategy());
    val sslsf = new SSLConnectionSocketFactory(sslCtxBuilder.build(), NoopHostnameVerifier.INSTANCE);
    builder.setSSLSocketFactory(sslsf)
  }

  private def readTimeout() = {
    Try(Integer.parseInt(System.getProperty("rest_timeout")))
      .getOrElse(100)
  }

  private def createBasicAuthHeader(user: String, password: String) = {
    val authenticate = BasicScheme.authenticate(new UsernamePasswordCredentials(user, password), "UTF-8", false)
    new BasicHeader("Authorization", authenticate.getValue)
  }

  private def buildUri(path: String, queryParams: Map[String, String]) = {
    val uriBuilder = new URIBuilder()
      .setScheme(if (ssl) "https" else "http")
      .setHost(host)
      .setPort(port)
      .setPath(("/" + path + "/").replaceAll("//", "/"))
    queryParams
      .foldLeft(uriBuilder) {
        case (builder, (name, value)) => builder.setParameter(name, value)
      }
      .build()
  }
}

object RestClient {

  def bodyFrom(response: HttpResponse): String = {
    EntityUtils.toString(response.getEntity)
  }
}